Raven.Client.Lightweight/Shard/AsyncShardedDocumentSession.cs

```diff
@@ -805,5 +806,53 @@ public Task<ResponseTimeInformation> ExecuteAllPendingLazyOperationsAsync(Cancel
             throw new InvalidOperationException("Document '" + documentKey + "' no longer exists and was probably deleted");
              
         }
+
+        public async Task<Operation> DeleteByIndexAsync<T, TIndexCreator>(Expression<Func<T, bool>> expression) where TIndexCreator : AbstractIndexCreationTask, new()
+        {
+            var indexCreator = new TIndexCreator();
+            return await DeleteByIndexAsync(indexCreator.IndexName, expression).ConfigureAwait(false);
+        }
+
+        private IRavenQueryable<T> QueryInternal<T>(IAsyncDatabaseCommands assyncDatabaseCommands, string indexName, bool isMapReduce = false)
+        {
+            var ravenQueryStatistics = new RavenQueryStatistics();
+            var highlightings = new RavenQueryHighlightings();
+            var ravenQueryInspector = new RavenQueryInspector<T>();
+            var ravenQueryProvider = new RavenQueryProvider<T>(this, indexName, ravenQueryStatistics, highlightings, null, assyncDatabaseCommands, isMapReduce);
+            ravenQueryInspector.Init(ravenQueryProvider,
+                ravenQueryStatistics,
+                highlightings,
+                indexName,
+                null,
+                this, null, assyncDatabaseCommands, isMapReduce);
+            return ravenQueryInspector;
+        }
+
+        public async Task<Operation> DeleteByIndexAsync<T>(string indexName, Expression<Func<T, bool>> expression)
+        {
+            var shards = GetCommandsToOperateOn(new ShardRequestData
+            {
+                EntityType = typeof(T),
+                Keys = { indexName }
+            });
+
+            var operations = shardStrategy.ShardAccessStrategy.ApplyAsync(shards, new ShardRequestData
+            {
+                EntityType = typeof(T),
+                Keys = { indexName }
+            }, (dbCmd, i) =>
+            {
+                var query = QueryInternal<T>(dbCmd, indexName).Where(expression);
+                var indexQuery = new IndexQuery()
+                {
+                    Query = query.ToString()
+                };
+
+                return dbCmd.DeleteByIndexAsync(indexName, indexQuery);
+            });
+
+            var result = await operations.ConfigureAwait(false);
+            return result[0]; // TODO: Return meaningful result
```

Comment:
Maybe create ShardedOperation that inherits from Operation and override WaitForCompletionAsync so it will wait for all operations on all shards?